server/config/service.go

```diff
@@ -114,45 +114,53 @@ func (s *service) Reconfigure(stored StoredConfig, services ...Configurable) err
 	mmconf := s.reloadMattermostConfig()
 	newConfig := s.Get()
 
+	s.log.Debugf("Reconfigure apps: %v installed, %v listed",
+		len(stored.InstalledApps), len(stored.LocalManifests))
+
 	// GetLicense silently drops an RPC error
 	// (https://github.com/mattermost/mattermost-server/blob/fc75b72bbabf7fabfad24b9e1e4c321ca9b9b7f1/plugin/client_rpc_generated.go#L864).
 	// When running in Mattermost cloud we must not fall back to the on-prem mode, so in case we get a nil retry once.
 	license := s.mm.System.GetLicense()
 	if license == nil {
 		license = s.mm.System.GetLicense()
 		if license == nil {
-			s.log.Infof("Failed to fetch license twice. May incorrectly default to on-prem mode.")
+			s.log.Debugf("Failed to fetch license twice. May incorrectly default to on-prem mode")
 		}
 	}
 
-	err := newConfig.Update(stored, mmconf, license, s.log)
+	err := newConfig.update(stored, mmconf, license, s.log)
 	if err != nil {
+		s.log.WithError(err).Infof("failed to update config")
```

Comment:
Log _and_ return seems off. Why can't the caller log?

Reply:
@hanzei 0/5 I'd argue that _in general_ errors should be logged where they happen, and the logger used should be the most contextualized.

But I also agree with your statement, and the redundancy of the message bothers me. I tried to untangle this to fix in its entirety, there are other instances of the "log and return" (anti?-)pattern. 1/5 a separate ticket to rework the logging, add consistency/best practices?